openzwavemqtt/util/mqtt_client.py

```diff
@@ -0,0 +1,262 @@
+        try:
+            await self.publish(*args, **kwargs)
+        except MqttError as err:
+            LOGGER.error("Failed sending message: %s", err)
+
+    def add_manager(self, manager: OZWManager) -> None:
+        """Add an OZW manager."""
+        self.managers[id(manager)] = manager
+
+    def send_message(self, topic: str, payload: Union[str, dict]) -> None:
+        """Send a message from the manager options."""
+        task = asyncio.create_task(self.safe_publish(topic, json.dumps(payload)))
+        self.tasks.add(task)
+
+    async def start_client(self) -> None:
+        """Start the client."""
+        self.client_task = asyncio.current_task()
+        # Reconnect automatically until the client is stopped.
+        while True:
+            try:
+                await self.subscribe_managers()
+            except MqttError as err:
+                self.reconnect_interval = min(self.reconnect_interval * 2, 900)
+                LOGGER.error(
+                    "MQTT error: %s. Reconnecting in %s seconds",
+                    err,
+                    self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+                self.create_client()  # reset connect/reconnect futures
+
+    async def subscribe_managers(self) -> None:
+        """Connect and subscribe to manager topics."""
+        async with AsyncExitStack() as stack:
+            # Keep track of the asyncio tasks that we create, so that
+            # we can cancel them on exit.
+            tasks = self.tasks = set()
+            stack.push_async_callback(cancel_tasks, tasks)
+
+            # Connect to the MQTT broker.
+            await stack.enter_async_context(self.asyncio_client)
+            # Reset the reconnect interval after successful connection.
+            self.reconnect_interval = 1
+
+            # Messages that doesn't match a filter will get logged and handled here.
+            messages = await stack.enter_async_context(
+                self.asyncio_client.unfiltered_messages()
+            )
+
+            for manager in self.managers.values():
+                task = asyncio.create_task(
+                    handle_messages(messages, manager.receive_message)
+                )
+                tasks.add(task)
+
+                # Note that we subscribe *after* starting the message loggers.
+                # Otherwise, we may miss retained messages.
+                topic = f"{manager.options.topic_prefix}#"
+                await self.subscribe(topic)
+
+            # Wait for everything to complete (or fail due to, e.g., network errors).
+            # Make sure we await new tasks added while awaiting the first tasks.
+            while self.tasks:
+                current_tasks = list(self.tasks)
+                self.tasks.clear()
+                await asyncio.gather(*current_tasks)
+
+    async def unsubscribe_manager(self, manager: OZWManager) -> None:
```

Comment:
We should also cancel the manager receive message task. 

I wonder if it's easier to not create tasks for `handle_messages` but instead whenever you receive a message, check which manager matches and then send it to them.

Something like this

```python
async for message in messages:
    for manager in self.managers.values():
        if not matches(message.topic, manager.topic):
            continue

        _LOGGER.debug("Received message", message)
        manager.handle(message)
        break

    else:
        _LOGGER.warning("Received unhandled message")
```

Reply:
That will make it a lot easier to track what's going on instead of juggling tasks. It also means we don't need to deal with managers being registered after the client is up.

Reply:
Actually thinking about it… we should maybe not support multiple managers at all ? Keep it simple. 

The OZW add-on can only run one instance anyway?

Reply:
Yes. But we may want to support more than one instance per add-on in the future?

We can drop it for now though to keep it simple. Build it when we need it.

Reply:
To support two instances, we would also have to support 2 USB sticks. Let's keep it for the future.

Reply:
We'll still need at least one task that publishes messages, I think.

Reply:
Correction, we need two tasks, one for publish and one for incoming messages, since we don't want to subscribe until after we start the incoming message handler.

Reply:
It doesn't look like publish requires a permanent task? It's short lived tasks that pass the message on. The main task of the client doesn't have to take care of that?

Reply:
I changed it so that it's up to the user of the client to pass in the correct manager when starting and unsubscribing now. What do you think?